File: src/game/systems/falldeath.lua
local Aspect = require 'ecs.aspect'
local System = require 'ecs.system'
local Fixture = require 'game.components.fixture'
local Health = require 'game.components.health'
local Player = require 'game.components.player'
local Position = require 'game.components.position'

local aspect = Aspect.new({Fixture, Health, Player, Position})
local Death = System:new('falldeath', aspect)

local BUFFER = 100

function Death:update(dt)
  for _, entity in pairs(self.entities) do
    local position = entity:as(Position)

    if position.y > WINDOW_HEIGHT + BUFFER then
      entity:as(Health).hitpoints = 0
    end
  end
end

return Death

File: client2/src/game/systems/FallDeath.ts
import { WINDOW_HEIGHT } from '../../conf'
import { System } from '../../ecs/System'
import { Aspect } from '../../ecs/Aspect'
import { SystemFlag } from '../flags'
import { GameObject } from '../components/GameObject'
import { Health } from '../components/Health'
import { Player } from '../components/Player'
import { Position } from '../components/Position'

const BUFFER = 100

export class FallDeathSystem extends System {
  static _id = 'FallDeath'
  _id = FallDeathSystem._id

  static _flag = SystemFlag.FallDeath
  _flag = FallDeathSystem._flag

  static _aspect = new Aspect([GameObject, Health, Player, Position])
  _aspect = FallDeathSystem._aspect

  update = (dt: number) => {
    this.entities.forEach(entity => {
      const position = entity.as(Position)

      if (position && position.y > WINDOW_HEIGHT + BUFFER) {
        const health = entity.as(Health)

        if (health) {
          health.hitpoints = 0
        }
      }
    })
  }
}
